### ISSUE TITLE
No widevine components for 16.x

### ISSUE THREAD

**Issue by ArmelChesnais** [castlabs/electron-releases#116]
I'm currently setting up to upgrade my local version of Castlabs electron from 13.5.0 to 16.0.2 (some vanilla Electron issues prevented 14.x and 15.x upgrade so far), but it appears that widevine components are not getting installed, or not being detected in the browser? I have run the signing process as well.

```Javascript
console.log(`before whenReady`)
app.whenReady().then(async () => {
  console.log(`whenReady`)
  const t0 = performance.now()
  await components.whenReady().catch(err => console.log(`component ready fail:`, err));
  console.log('components ready:', components.status());
  console.log('components performance:', performance.now() - t0);
  createWindow();
})
```
Displays this in the terminal:
```
before whenReady
whenReady
components ready: null
components performance: 0.5515000820159912
createWindow
```
Additionally, navigating in the app to `https://bitmovin.com/demos/drm`
Displays that Widevine is not present (comparatively to running 13.5.0 which does display it as present): 
![image](https://user-images.githubusercontent.com/23585954/144149795-55ed3810-3e83-4158-8acd-2d5a32c2933d.png)

The Wiki appears to display 16.x as in development with widevine CDM TBD. 
Is this accurate, and therefore why there is no widevine component at this time for this version? Or is this out of date since Castlabs' electron 16.x left beta, and something else is not functioning?

**Comment by khwaaj** [comment]
Which platform is this? I just verified that this works fine on an Intel Mac using the `electron-v16.0.2+wvcus-darwin-x64` release build, using the exact snippet of code you are using above.

The TBD on the Wiki page is just an oversight, we simply forgot to update it when v16 went stable. I'll make sure to fix it, thanks for reporting it.

**Comment by khwaaj** [comment]
Btw, your output should look something like this.

First run (with no CDM installed):
```
before whenReady
whenReady
components ready: {
  oimompecagnajdejgnnjijobebaeigek: {
    name: 'Widevine Content Decryption Module',
    status: 'updated',
    version: '4.10.2391.0'
  }
}
components performance: 790.0901508331299
```

Second run:
```
before whenReady
whenReady
components ready: {
  oimompecagnajdejgnnjijobebaeigek: {
    name: 'Widevine Content Decryption Module',
    status: 'new',
    version: '4.10.2391.0'
  }
}
components performance: 10.695983171463013
```

**Comment by khwaaj** [comment]
Come to think of it, the issue you describe is what would happen if the `components.whenReady()` call comes too early, before the updater is initialized. I'm not sure why that would happen here, but there could be some kind of race condition in the new ComponentUpdater code causing this. If so, introducing a delay before the call should make it start working.

**Comment by ArmelChesnais** [comment]
This is on Windows 10, sorry, should have specified in the first post. 

Will check putting in a delay when I have time to test it out. Hopefully a fix can be put it, as I'm not too fond of introducing magic numbers in.

**Comment by ArmelChesnais** [comment]
I've attempted to add a timeout, but still seeing the same results
```Javascript
console.log(`before whenReady`)
app.whenReady().then(async () => {
  console.log(`whenReady`)
  await new Promise((resolve) => { setTimeout(resolve, 1000) })
  const t0 = performance.now()
  await components.whenReady().catch(err => console.log(`component ready fail:`, err));
  console.log('components ready:', components.status());
  console.log('components performance:', performance.now() - t0);
  createWindow();
})
```

**Comment by khwaaj** [comment]
Ok, thanks for the feedback.

The timeout was just a suggestion for a temporary workaround and/or verification of the issue. Now that I know the platform I'll see about reproducing it so that it can be fixed for real.

**Comment by khwaaj** [comment]
Weirdly, I cannot reproduce this on my Windows 10 machine using a `x64` build of `v16.0.2+wvcus`. Even introducing artificial delays in the ECS code to try and force a condition where the updater is not initialized before `app.whenReady()` fails to do so.

Could you share some more details around your setup & implementation?
* Is this just a minimal implementation like the README example? If not, can you reproduce with a minimal implementation?
* How are you starting the app? 
* Are you passing any additional command line parameters?
* Are you setting any configuration properties?
* Have you tried clearing any old configuration, e.g. in `$HOME/AppData/Roaming/$APP-NAME` (you can just rename/move if you want to be able to restore it to test further)
* Is this running on a physical machine or a VM?

**Comment by ArmelChesnais** [comment]
- Issue was in our project, not minimal implementation. Tried minimal setup just now, and it does indeed load the widevine CDM correctly.
- our setup is based on an older version of the electron-react-boilerplate. We use this script to start the app: `"start": "cross-env NODE_ENV=production electron ./app/main.prod.js",`
- we do pass command line parameters, but I've also tried commenting them all off, and saw no change.
- will have to check for configuration properties, but not sure we have anything other than the command line that I can tell so far. We have a number of modules being used that may tap into electron, so I'd have to dig through to see what's going on there.
- tried removing the widevine folder in Roaming, but saw no change in behaviour. Haven't removed the whole folder. Though the minimal setup and our dev environment both point to `$HOME/AppData/Roaming/Electron`, and as mentioned above, the minimal setup did work as expected.
- this is a physical machine

This is happening in our dev environment, can try packaging as well to see if behaviour happens there too.

**Comment by ArmelChesnais** [comment]
Also as a quick update, it appears to happen on Mac too. Will try to get more details when I have some time.

**Comment by ArmelChesnais** [comment]
After reducing our main file to the barest elements, and still getting the same behaviour of missing widevine:
```Javascript
import {
  app,
  components,
} from 'electron';

const createWindow = async () => {
  console.log(`createWindow`)
  return
}

console.log('before whenReady')
app.whenReady().then(async () => {
  console.log('in whenReady')

  await new Promise(resolve => setTimeout(resolve, 1000))
  const t0 = performance.now()
  await components.whenReady();
  console.log('components ready:', components.status());
  console.log('components performance:', performance.now() - t0);
  createWindow();
})
```

APPARENT SOLUTION:
I tried bypassing babel/webpack and using a raw .js file, and had to convert my `import`s to `require` statements, and that seems to have solved it in all environments so far
```Javascript
const {
  app,
  components,
} = require('electron');
```
I'll update if I find any further breaks

**Comment by khwaaj** [comment]
Thanks for the feedback.

I'm not sure what to think, so I set up a minimal babel project to try it out myself. It looks like the combination of `import` and `async/await` causes problems in this case. Like you said, using `require` works fine in all cases, but using `import` together with `async/await` fails.

Changing the code to use `.then()` instead also makes it work, even with `import`, e.g:

```js
app.whenReady().then(async () => {
  await components.whenReady()
```

To:

```js
app.whenReady().then(components.whenReady()).then(() => {
```

**Comment by khwaaj** [comment]
I'm pretty sure I figured out what the problem is. The code generated in the failing case triggers an issue in our implementation which will cause an uninitialized instance of the components API to be used.

I'll see about fixing this on our side. In the mean-time one of the workarounds above can be used to avoid the problem. An alternative workaround would be to explicitly create an instance of `components` early and store it, e.g:

```js
import {app, components, BrowserWindow} from 'electron';
const cus = components;

app.whenReady().then(async () => {
  await cus.whenReady();
  console.log('components ready:', cus.status());

  ...
}
```

**Comment by khwaaj** [comment]
I just published [16.0.3+wvcus.1](https://github.com/castlabs/electron-releases/releases/tag/v16.0.3%2Bwvcus.1), which takes care of this issue. Please take it for a spin and let me know how it goes!

**Comment by khwaaj** [comment]
Closing due to inactivity. Feel free to reopen if issue persists.
